src/becca/entities/battachment.ts

```diff
@@ -1,28 +1,56 @@
 "use strict";
 
-const utils = require('../../services/utils.js');
-const dateUtils = require('../../services/date_utils.js');
-const AbstractBeccaEntity = require('./abstract_becca_entity.js');
-const sql = require('../../services/sql.js');
-const protectedSessionService = require('../../services/protected_session.js');
-const log = require('../../services/log.js');
+import utils = require('../../services/utils');
+import dateUtils = require('../../services/date_utils');
+import AbstractBeccaEntity = require('./abstract_becca_entity');
+import sql = require('../../services/sql');
+import protectedSessionService = require('../../services/protected_session');
+import log = require('../../services/log');
+import { AttachmentRow } from './rows';
+import BNote = require('./bnote');
+import BBranch = require('./bbranch');
 
 const attachmentRoleToNoteTypeMapping = {
     'image': 'image'
 };
 
+interface ContentOpts {
+    // TODO: Found in bnote.ts, to check if it's actually used and not a typo.
+    forceSave?: boolean;
+
+    /** will also save this BAttachment entity */
+    forceFullSave?: boolean;
+    /** override frontend heuristics on when to reload, instruct to reload */
+    forceFrontendReload?: boolean;
+}
+
 /**
  * Attachment represent data related/attached to the note. Conceptually similar to attributes, but intended for
  * larger amounts of data and generally not accessible to the user.
- *
- * @extends AbstractBeccaEntity
  */
-class BAttachment extends AbstractBeccaEntity {
+class BAttachment extends AbstractBeccaEntity<BAttachment> {
     static get entityName() { return "attachments"; }
     static get primaryKeyName() { return "attachmentId"; }
     static get hashedProperties() { return ["attachmentId", "ownerId", "role", "mime", "title", "blobId", "utcDateScheduledForErasureSince"]; }
 
-    constructor(row) {
+    noteId?: number;
```

Comment:
@ProfDoof , `noteId` is only used in a few places in `battachment`. It might actually be a bug.

The parent of an attachment is not `noteId`, but `ownerId` which the comments said that it's either a  `noteId` or a `revisionId`.

`noteId` seems to be used during the conversion of an attachment to a note only, and it's not present in the constructor so that's why I had to make it optional.

Reply:
Interesting. Hmm, in that case, in the future it might be best to get rid of the extra IDs and just provide the owner id and possibly a discriminant to know whether we have a revision attachment or an active note attachment. But outside the scope here.